Release/src/http/client/http_client_asio.cpp

```diff
@@ -325,32 +325,64 @@ class asio_connection_pool
     boost::asio::io_service& m_io_service;
     const int m_timeout_secs;
     const bool m_start_with_ssl;
-    const std::function<void(boost::asio::ssl::context&)>& m_ssl_context_callback;
+    std::function<void(boost::asio::ssl::context&)> m_ssl_context_callback;
     std::vector<std::shared_ptr<asio_connection> > m_connections;
     std::mutex m_connections_mutex;
 };
 
 
-
 class asio_client : public _http_client_communicator
 {
 public:
     asio_client(http::uri address, http_client_config client_config)
     : _http_client_communicator(std::move(address), std::move(client_config))
-    , m_pool(crossplat::threadpool::shared_instance().service(),
-             base_uri().scheme() == "https" && !_http_client_communicator::client_config().proxy().is_specified(),
-             std::chrono::seconds(30), // Unused sockets are kept in pool for 30 seconds.
-             this->client_config().get_ssl_context_callback()) 
     , m_resolver(crossplat::threadpool::shared_instance().service())
-    {}
+    {
+        if (this->client_config().get_ssl_context_callback())
+        {
+            // The pool is not added to the map because there is no better approaches to compare callback functors.
+            m_pool = std::make_shared<asio_connection_pool>(crossplat::threadpool::shared_instance().service(),
+                base_uri().scheme() == "https" && !_http_client_communicator::client_config().proxy().is_specified(),
+                std::chrono::seconds(30), // Unused sockets are kept in pool for 30 seconds.
+                this->client_config().get_ssl_context_callback());
+        }
+        else
+        {
+            std::string host = base_uri().to_string();
+
+            auto &credentials = _http_client_communicator::client_config().credentials();
+            if (credentials.is_set())
+            {
+                host.append(credentials.username());
+            }
+
+            auto &proxy = _http_client_communicator::client_config().proxy();
+            if (proxy.is_specified())
+            {
+                host.append(proxy.address().to_string());
+                if (proxy.credentials().is_set())
+                {
+                  host.append(proxy.credentials().username());
+                }
+            }
+
+            m_pool = crossplat::threadpool::shared_instance().obtain_connection_pool(host, [this]()
+            {
+                return std::make_shared<asio_connection_pool>(crossplat::threadpool::shared_instance().service(),
+                    base_uri().scheme() == "https" && !_http_client_communicator::client_config().proxy().is_specified(),
+                    std::chrono::seconds(30), // Unused sockets are kept in pool for 30 seconds.
+                    this->client_config().get_ssl_context_callback());
```

Comment:
This branch of the `if` already has guaranteed there is no ssl_context_callback, so passing in `nullptr` here is probably more clear.